soroban-spec/src/gen/ts.rs

```diff
@@ -0,0 +1,351 @@
+    } else {
+        let doc = doc.split('\n').join("\n * ");
+        format!(
+            r#"/**
+ * {doc}
+ */
+"#,
+        )
+    }
+}
+
+pub fn entry_to_ts(entry: &Entry) -> String {
+    match entry {
+        Entry::Function {
+            doc,
+            name,
+            inputs,
+            outputs,
+        } => {
+            let args = inputs
+                .iter()
+                .map(|i| format!("((i) => {})({})", type_to_js_xdr(&i.value), i.name))
+                .join(",\n        ");
+            let input = (!inputs.is_empty())
+                .then(|| {
+                    format!(
+                        "{{{}}}: {{{}}}",
+                        inputs.iter().map(func_input_to_arg_name).join(", "),
+                        inputs.iter().map(func_input_to_ts).join(", ")
+                    )
+                })
+                .unwrap_or_default();
+            let mut is_result = false;
+            let mut inner_return_type = String::new();
+            let return_type = if outputs.is_empty() {
+                "".to_owned()
+            } else if outputs.len() == 1 {
+                inner_return_type = type_to_ts(&outputs[0]);
+                is_result = inner_return_type.starts_with("Result<");
+                format!(": Promise<{inner_return_type}>")
+            } else {
+                format!(
+                    ": Promise<[{}]>>",
+                    outputs.iter().map(type_to_ts).join(", ")
+                )
+            };
+            let ts_doc = doc_to_ts_doc(doc);
+
+            // let output_parser = outputs.get(0).map(scVal_to_type).unwrap_or_default();
+            if is_result {
+                inner_return_type = inner_return_type
+                    .strip_prefix("Result<")
+                    .unwrap()
+                    .strip_suffix('>')
+                    .unwrap()
+                    .to_owned();
+            }
+
+            let mut output = (!outputs.is_empty())
+                .then(|| format!("scValStrToJs(response.xdr) as {inner_return_type}"))
+                .unwrap_or_default();
+            if is_result {
+                output = format!("new Ok({output})");
+            }
+            let mut output = format!(
+                r#"
+    // @ts-ignore Type does exist
+    const response = await invoke(invokeArgs);
+    return {output};
+    "#
+            );
+            if is_result {
+                output = format!(
+                    r#"
+    try {{
+        {output}
+    }} catch (e) {{
+        //@ts-ignore
+        let err = get_error(e.message);
+        if (err) {{
+            return err;
+        }} else {{
+            throw e;
+        }}
+    }}"#
+                );
+            }
+            let args = (!inputs.is_empty())
+                .then(|| format!("args: [{args}], "))
+                .unwrap_or_default();
+            format!(
+                r#"{ts_doc}export async function {name}({input}){return_type} {{
```

Comment:
Might be able to clean up a bit of code above by always returning a Promise, but for no result: `Promise<void>`

Reply:
Good idea!